File: 207.course-schedule.cpp
/*
 * @lc app=leetcode id=207 lang=cpp
 *
 * [207] Course Schedule
 */

// @lc code=start
class Solution {
 public:
  enum Color { white, grey, black };
  bool canFinish(int numCourses, vector<vector<int>> &prerequisites) {
    vector<vector<int>> graph(numCourses);
    vector<Color> visited(numCourses, white);
    for (auto &p : prerequisites) {
      graph[p[1]].push_back(p[0]);
    }

    for (int i = 0; i < numCourses; ++i) {
      if (visited[i] == white and !dfs(graph, i, visited)) {
        return false;
      }
    }
    return true;
  }

  bool dfs(vector<vector<int>> &graph, int i, vector<Color> &visited) {
    visited[i] = grey;
    for (auto n : graph[i]) {
      if (visited[n] == grey) {
        return false;
      } else if (visited[n] == white and !dfs(graph, n, visited)) {
        return false;
      }
    }
    visited[i] = black;
    return true;
  }
};
// @lc code=end

File: 207.course-schedule.py
#
# @lc app=leetcode id=207 lang=python3
#
# [207] Course Schedule
#

# @lc code=start
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # no circle
        graph = [[] for _ in range(numCourses)]
        for edge in prerequisites:
            graph[edge[0]].append(edge[1])

        state = [0 for _ in range(numCourses)]
        
        def dfs(node: int) -> bool:
            state[node] = 1
            for t in graph[node]:
                if state[t] == 1:
                    return False
                if state[t] == 0 and not dfs(t):
                    return False
            state[node] = 2
            return True
        
        for i in range(numCourses):
            if state[i] == 0 and not dfs(i):
                return False
        
        return True

                    


            


# @lc code=end

